src/ln/channelmanager.rs

```diff
@@ -2020,26 +2021,41 @@ impl ChannelManager {
 
 	// Process failure we got back from upstream on a payment we sent. Returns update and a boolean
 	// indicating that the payment itself failed
-	fn process_onion_failure(&self, htlc_source: &HTLCSource, mut packet_decrypted: Vec<u8>) -> (Option<msgs::HTLCFailChannelUpdate>, bool) {
+	fn process_onion_failure(&self, htlc_source: &HTLCSource, mut packet_decrypted: Vec<u8>) -> (Option<msgs::HTLCFailChannelUpdate>, bool, Option<u16>) {
 		if let &HTLCSource::OutboundRoute { ref route, ref session_priv, ref first_hop_htlc_msat } = htlc_source {
 			macro_rules! onion_failure_log {
-				( $error_code_textual: expr, $error_code: expr, $reported_name: expr, $reported_value: expr ) => {
-					log_trace!(self, "{}({:#x}) {}({})", $error_code_textual, $error_code, $reported_name, $reported_value);
-				};
-				( $error_code_textual: expr, $error_code: expr ) => {
-					log_trace!(self, "{}({})", $error_code_textual, $error_code);
-				};
+				( $error_code_textual: expr, $error_code: expr, $reported_name: expr, $reported_value: expr ) => {{
+					log_info!(self, "{}({:#x}) {}({})", $error_code_textual, $error_code, $reported_name, $reported_value);
+				}};
+				( $error_code_textual: expr, $error_code: expr ) => {{
+					log_info!(self, "{}({:#x})", $error_code_textual, $error_code);
+				}};
+			}
+
+			// when a node has sent invalid code or malformed failure msg
+			macro_rules! return_node_fail_update {
+				($res:ident, $node_id:expr, $retry:expr) => {{
+					$res = Some((Some(msgs::HTLCFailChannelUpdate::NodeFailure {
+						node_id: $node_id,
+						is_permanent: true,
+					}), $retry));
+					return;
+				}}
 			}
 
 			const BADONION: u16 = 0x8000;
 			const PERM: u16 = 0x4000;
+			const NODE: u16 = 0x2000;
```

Comment:
I think we can constify all error codes, already quoted in #49 ?

Reply:
ah.. if we introduce constant error tables, it may as well to flatten out process_onion's error_code branching.